op-e2e/faultproofs/precompile_test.go

```diff
@@ -0,0 +1,183 @@
+			log.Info("genesis", "l2", sys.RollupConfig.Genesis.L2, "l1", sys.RollupConfig.Genesis.L1, "l2_time", sys.RollupConfig.Genesis.L2Time)
+
+			l1Client := sys.Clients["l1"]
+			l2Seq := sys.Clients["sequencer"]
+			rollupRPCClient, err := rpc.DialContext(context.Background(), sys.RollupNodes["sequencer"].HTTPEndpoint())
+			require.Nil(t, err)
+			rollupClient := sources.NewRollupClient(client.NewBaseRPCClient(rollupRPCClient))
+
+			aliceKey := cfg.Secrets.Alice
+
+			t.Log("Capture current L2 head as agreed starting point")
+			latestBlock, err := l2Seq.BlockByNumber(ctx, nil)
+			require.NoError(t, err)
+			agreedL2Output, err := rollupClient.OutputAtBlock(ctx, latestBlock.NumberU64())
+			require.NoError(t, err, "could not retrieve l2 agreed block")
+			l2Head := agreedL2Output.BlockRef.Hash
+			l2OutputRoot := agreedL2Output.OutputRoot
+
+			receipt := op_e2e.SendL2Tx(t, cfg, l2Seq, aliceKey, func(opts *op_e2e.TxOpts) {
+				opts.Gas = 1_000_000
+				opts.ToAddr = &test.address
+				opts.Nonce = 0
+				opts.Data = test.input
+			})
+
+			t.Log("Determine L2 claim")
+			l2ClaimBlockNumber := receipt.BlockNumber
+			l2Output, err := rollupClient.OutputAtBlock(ctx, l2ClaimBlockNumber.Uint64())
+			require.NoError(t, err, "could not get expected output")
+			l2Claim := l2Output.OutputRoot
+
+			t.Log("Determine L1 head that includes all batches required for L2 claim block")
+			require.NoError(t, waitForSafeHead(ctx, l2ClaimBlockNumber.Uint64(), rollupClient))
+			l1HeadBlock, err := l1Client.BlockByNumber(ctx, nil)
+			require.NoError(t, err, "get l1 head block")
+			l1Head := l1HeadBlock.Hash()
+
+			inputs := cannon.LocalGameInputs{
+				L1Head:        l1Head,
+				L2Head:        l2Head,
+				L2Claim:       common.Hash(l2Claim),
+				L2OutputRoot:  common.Hash(l2OutputRoot),
+				L2BlockNumber: l2ClaimBlockNumber,
+			}
+			runCannon(t, ctx, sys, inputs, "sequencer")
+		})
+	}
+}
+
+func runCannon(t *testing.T, ctx context.Context, sys *op_e2e.System, inputs cannon.LocalGameInputs, l2Node string) {
+	l1Endpoint := sys.NodeEndpoint("l1")
+	l1Beacon := sys.L1BeaconEndpoint()
+	cannonOpts := challenger.WithCannon(t, sys.RollupCfg(), sys.L2Genesis(), sys.RollupEndpoint(l2Node), sys.NodeEndpoint(l2Node))
+	dir := t.TempDir()
+	proofsDir := filepath.Join(dir, "cannon-proofs")
+	cfg := config.NewConfig(common.Address{}, l1Endpoint, l1Beacon, dir)
+	cannonOpts(&cfg)
+
+	logger := testlog.Logger(t, log.LevelInfo).New("role", "cannon")
+	executor := cannon.NewExecutor(logger, metrics.NoopMetrics, &cfg, inputs)
+
+	t.Log("Running cannon")
+	err := executor.GenerateProof(ctx, proofsDir, math.MaxUint)
+	require.NoError(t, err, "failed to generate proof")
+
+	state, err := parseState(filepath.Join(proofsDir, "final.json.gz"))
+	require.NoError(t, err, "failed to parse state")
+	t.Logf("Completed in %d steps", state.Step)
+}
+
+func waitForSafeHead(ctx context.Context, safeBlockNum uint64, rollupClient *sources.RollupClient) error {
```

Comment:
Can we make this a reusable function in the `wait` package like we have `wait.ForUnsafeBlock`? https://github.com/ethereum-optimism/optimism/blob/02987acc48daa480a085ac748062a760723b9e2d/op-e2e/e2eutils/wait/blocks.go#L75

Reply:
Done. til of wait/blocks.go. Very useful.